File: built-in plugins dsp/mb-creative-vocoder.cpp
/**
 * MB Vocoder
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic vocoder with adjustable band count and formant tracking
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CREATIVE_VOCODER_H
#define MB_CREATIVE_VOCODER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCreativeVocoder : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-creative-vocoder";
    static constexpr const char* PLUGIN_NAME    = "MB Vocoder";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bands = 16f;  // range [4, 32]
    float formant = 0f;  // range [-12, 12]
    float attack = 10f;  // range [1, 100]
    float release = 50f;  // range [5, 500]
    float mix = 1f;  // range [0, 1]
    };

    MbCreativeVocoder() = default;
    ~MbCreativeVocoder() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bands = std::clamp(params.bands, 4f, 32f);
        params.formant = std::clamp(params.formant, -12f, 12f);
        params.attack = std::clamp(params.attack, 1f, 100f);
        params.release = std::clamp(params.release, 5f, 500f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Vocoder
        return input;
    }
};

#endif // MB_CREATIVE_VOCODER_H

File: maxbooster7.5/built-in plugins dsp/mb-creative-vocoder.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbCreativeVocoderPlugin: PluginDefinition = {
  id: "mb-creative-vocoder",
  slug: "mb-creative-vocoder",
  name: "MB Vocoder",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description:
    "Classic vocoder with adjustable band count and formant tracking",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "bands",
      name: "Bands",
      type: "float",
      defaultValue: 16,
      minValue: 4,
      maxValue: 32,
      automatable: false,
    },
    {
      id: "formant",
      name: "Formant Shift",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "attack",
      name: "Attack",
      type: "float",
      defaultValue: 10,
      minValue: 1,
      maxValue: 100,
      automatable: true,
    },
    {
      id: "release",
      name: "Release",
      type: "float",
      defaultValue: 50,
      minValue: 5,
      maxValue: 500,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { bands: 16, formant: 0, attack: 10, release: 50, mix: 1 },
};

export default MbCreativeVocoderPlugin;
